components/fxa-client/src/migrator.rs

```diff
@@ -24,40 +26,72 @@ impl FirefoxAccount {
         k_sync: &str,
         k_xcs: &str,
         copy_session_token: bool,
-    ) -> Result<()> {
+    ) -> Result<bool> {
         // if there is already a session token on account, we error out.
         if self.state.session_token.is_some() {
             return Err(ErrorKind::IllegalState("Session Token is already set.").into());
         }
 
-        let migration_session_token = if copy_session_token {
-            let duplicate_session = self
-                .client
-                .duplicate_session(&self.state.config, &session_token)?;
+        self.state.in_flight_migration = Some(MigrationData {
+            k_sync: k_sync.to_string(),
+            k_xcs: k_xcs.to_string(),
+            copy_session_token,
+            session_token: session_token.to_string(),
+        });
+
+        let client = self.client.clone();
+        match self.helper_migration_network_methods(client) {
+            Ok(_) => {
+                log::info!("FxA migration complete");
+                Ok(true)
+            }
+            Err(err) => {
+                log::info!("Failed to perform network requests: {}", err);
+                // don't throw hard error here, let the consumers try again
+                Ok(false)
+            }
+        }
+    }
+
+    pub fn helper_migration_network_methods(&mut self, client: Arc<dyn FxAClient>) -> Result<()> {
```

Comment:
There are two ways that this could fail, and I think they need to be treated differently.

In one case, it might fail due to some sort of transient network error. That's fine, trying again later will probably resolve the issue.

In the other case, it might fail due to some sort of permanently-fatal error. For example the provided `sessionToken` might be invalid, and we don't find that out until we hit the network and get an auth error back from the server. We must avoid getting stuck in an endless retry loop in the case of such permanent errors.

I think it's fine to re-throw the error in both cases, but you're going to need to clear the in-flight migration state in the case of a permanent error.